### PR TITLE
[Pladform] Fix redirection to external player

### PR DESCRIPTION
**Pull request by KiberInfinity**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

Fix external player redirect URL extraction. Examples:
http://out.pladform.ru/player?pl=18079&type=html5&videoid=100246529 -> rutube
http://out.pladform.ru/player?pl=18079&type=html5&videoid=100231282 -> rutube

### CODE REVIEW

File: yt_dlp/extractor/pladform.py [hunk 1]
```diff
@@ -63,13 +63,20 @@ def _real_extract(self, url):
...
-            })
+            }, fatal=False)
 
         def fail(text):
             raise ExtractorError(
                 '%s returned error: %s' % (self.IE_NAME, text),
                 expected=True)
 
+        if not video:
+            targetUrl = self._request_webpage(url, None, note='Resolving final URL').geturl()
```
Comment (pukkandan) [inline yt_dlp/extractor/pladform.py:74]:
```suggestion
            targetUrl = self._request_webpage(HEADRequest(url), video_id, note='Resolving final URL').geturl()
```

File: yt_dlp/extractor/pladform.py [hunk 2]
```diff
@@ -63,13 +63,20 @@ def _real_extract(self, url):
...
+            if (targetUrl != url):
+                return self.url_result(targetUrl)
+            else:
+                fail("Can't parse page")
```
Comment (pukkandan) [inline yt_dlp/extractor/pladform.py:78]:
```py
            if targetUrl == url:
                raise ExtractorError('Can\'t parse page')
            return self.url_result(targetUrl)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
Also add/update tests

File: yt_dlp/extractor/pladform.py (hunk 1 above)
Comment (KiberInfinity) [reply yt_dlp/extractor/pladform.py:74]: HEAD request not allowed in tested cases

Comment (pukkandan) [reply yt_dlp/extractor/pladform.py]: You can revert that part then. But add the video_id

Comment (KiberInfinity) [reply yt_dlp/extractor/pladform.py]: Already done in latest commit

**Comment by pukkandan** [comment]
> Also add/update tests

**Comment by KiberInfinity** [comment]
> Also add/update tests

```cmd
python -m yt_dlp --verbose "http://out.pladform.ru/player?pl=18079&type=html5&videoid=100231282"
```
Worked successuly. Log:
<details><summary>Click to expand</summary><p>  

```
[debug] Command-line config: ['--verbose', 'http://out.pladform.ru/player?pl=18079&type=html5&videoid=100231282']  
[debug] Encodings: locale cp1251, fs utf-8, out utf-8, err utf-8, pref cp1251
[debug] yt-dlp version 2022.01.21 [f20d607b0] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: bc25ffc3b
[debug] Python version 3.8.5 (CPython 32bit) - Windows-10-10.0.19041-SP0
[debug] exe versions: ffmpeg 4.2.3
[debug] Optional libraries: Cryptodome, sqlite
[debug] Proxy map: {}
[debug] [Pladform] Extracting URL: http://out.pladform.ru/player?pl=18079&type=html5&videoid=100231282
[Pladform] 100231282: Downloading XML
WARNING: [Pladform] Unable to download XML: HTTP Error 404: Not Found
[Pladform] 100231282: Resolving final URL
[debug] [rutube] Extracting URL: https://rutube.ru/play/embed/6216d548e755edae6e8280667d774791/?event_source=pladform&pladform_player_id=18079&pladform_id=100231282&redirectref=
[rutube] 6216d548e755edae6e8280667d774791: Downloading video JSON
[rutube] 6216d548e755edae6e8280667d774791: Downloading options JSON
[rutube] 6216d548e755edae6e8280667d774791: Downloading f4m manifest
[rutube] 6216d548e755edae6e8280667d774791: Downloading f4m manifest
WARNING: [rutube] Unable to download f4m manifest: HTTP Error 404: Not Found
[rutube] 6216d548e755edae6e8280667d774791: Downloading f4m manifest
WARNING: [rutube] Unable to download f4m manifest: HTTP Error 404: Not Found
[rutube] 6216d548e755edae6e8280667d774791: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 6216d548e755edae6e8280667d774791: Downloading 1 format(s): m3u8-1298
[debug] Invoking downloader on "http://salam-1.rutube.ru/dive/video-1-304.rutube.ru/VPba4DhUOmHezl-3yXZ82g/hls-vod/EUFkkbM5bzDwoLjR2nDK6g/1643689289/124/0x500003970ba02e6c/676cca6081b44a7ea5c725ab9bc611d8.mp4.m3u8?i=576x432_1298"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 46
[download] Destination: Гарик Мартиросян и Гарик Харламов - Кастинг на концерт ко Дню милиции [6216d548e755edae6e8280667d774791].mp4
[download] 100% of 59.40MiB in 00:19
WARNING: Unable to extract metadata: ffprobe not found. Please install or provide the path using --ffmpeg-location
[FixupM3u8] Fixing MPEG-TS in MP4 container of "Гарик Мартиросян и Гарик Харламов - Кастинг на концерт ко Дню милиции [6216d548e755edae6e8280667d774791].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:Гарик Мартиросян и Гарик Харламов - Кастинг на концерт ко Дню милиции [6216d548e755edae6e8280667d774791].mp4" -map 0 -dn -ignore_unknown -c copy -f mp4 "-bsf:a" aac_adtstoasc -movflags "+faststart" "file:Гарик Мартиросян и Гарик Харламов - Кастинг на концерт ко Дню милиции [6216d548e755edae6e8280667d774791].temp.mp4"
```

</p></details>


But test:
```cmd
python test/test_download.py TestDownload.test_Pladform
```
always failed on non fatal errors:
<details><summary>Click to expand</summary><p>  


```
Traceback (most recent call last):
  File "test/test_download.py", line 170, in test_template
    res_dict = ydl.extract_info(
  File "C:\Dev\Work\udl\yt-dlp-ki\yt_dlp\YoutubeDL.py", line 1365, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "C:\Dev\Work\udl\yt-dlp-ki\yt_dlp\YoutubeDL.py", line 1392, in wrapper
    self.report_error(str(e), e.format_traceback())
  File "C:\Dev\Work\udl\yt-dlp-ki\yt_dlp\YoutubeDL.py", line 930, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "C:\Dev\Work\udl\yt-dlp-ki\yt_dlp\YoutubeDL.py", line 871, in trouble
    raise DownloadError(message, exc_info)
yt_dlp.utils.DownloadError: ERROR: [Pladform] 100231282: [Pladform] Unable to download XML: HTTP Error 404: Not Found; please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.; please 
report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with 
the --verbose flag and include its complete output.

----------------------------------------------------------------------
Ran 1 test in 0.568s

FAILED (errors=1)
```
</p></details>



Test configuration:
```
    {
        'url': 'http://out.pladform.ru/player?pl=18079&type=html5&videoid=100231282',
        'info_dict': {
            'id': '6216d548e755edae6e8280667d774791',
            'ext': 'mp4',
        },
    }
```

**Comment by pukkandan** [comment]
Add `expected_warnings` to the test dict. Examples: https://github.com/yt-dlp/yt-dlp/search?q=expected_warnings&type=code

**Pull request merged**